lib/services/s3.js

```diff
@@ -306,6 +341,42 @@ AWS.util.update(AWS.S3.prototype, {
   },
 
   /**
+   * Updates httpRequest with region. If region is not provided, then
+   * the httpRequest will be updated based on httpRequest.region
+   *
+   * @api private
+   */
+  updateReqBucketRegion: function updateReqBucketRegion(request, region) {
+    var httpRequest = request.httpRequest;
+    if (typeof region === 'string' && region.length) {
+      httpRequest.region = region;
+    }
+    if (!httpRequest.endpoint.host.match(/s3(?!-accelerate).*\.amazonaws\.com$/)) {
+      return;
+    }
+    var service = request.service;
+    var s3Config = service.config;
+    var s3BucketEndpoint = s3Config.s3BucketEndpoint;
+    if (s3BucketEndpoint) {
+      delete s3Config.s3BucketEndpoint;
+    }
+    var newConfig = AWS.util.copy(s3Config);
+    delete newConfig.endpoint;
+    newConfig.region = httpRequest.region;
+
+    var endpoint = (new AWS.S3(newConfig)).endpoint;
+    httpRequest.endpoint = new AWS.Endpoint(endpoint);
+    service.populateURI(request);
+    s3Config.s3BucketEndpoint = s3BucketEndpoint;
+    httpRequest.headers.Host = httpRequest.endpoint.host;
+
+    if (request._asm.currentState === 'validate') {
+      request.removeListener('build', service.populateURI);
```

Comment:
Not only would it duplicate work, it would break the endpoint for virtual-hosted buckets because the bucket name gets populated twice, e.g. `mybucketname.mybucketname.s3.amazonaws.com`. This only is a problem for requests that have not been built yet (requests still on the `validate` step), since requests that are being retried go straight to the `sign` step and do not go through the `build` step anymore. This is also why `populateURI` needs to be called inside this function, because requests being retried need to update their endpoint but won't go through the `build` step.